python/pyspark/sql/functions.py

```diff
@@ -2868,6 +2869,15 @@ def pandas_udf(f=None, returnType=None, functionType=None):
        +----------+--------------+------------+
        |         8|      JOHN DOE|          22|
        +----------+--------------+------------+
+       >>> @pandas_udf("first string, last string")  # doctest: +SKIP
```

Comment:
This is a nice feature!

I am wondering if we have a better way to handle the incorrect inputs. For example, if our end users specify incorrect return data types (e.g.  `@pandas_udf("first string, last int")`), do we issue a user-friend error message?

```
19/03/21 00:04:08 ERROR Executor: Exception in task 0.0 in stage 8.0 (TID 8)
org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/Users/xiaoli/IdeaProjects/sparkDelivery/python/lib/pyspark.zip/pyspark/worker.py", line 433, in main
    process()
  File "/Users/xiaoli/IdeaProjects/sparkDelivery/python/lib/pyspark.zip/pyspark/worker.py", line 428, in process
    serializer.dump_stream(func(split_index, iterator), outfile)
  File "/Users/xiaoli/IdeaProjects/sparkDelivery/python/lib/pyspark.zip/pyspark/serializers.py", line 360, in dump_stream
    self._assign_cols_by_name)
  File "/Users/xiaoli/IdeaProjects/sparkDelivery/python/lib/pyspark.zip/pyspark/serializers.py", line 316, in _create_batch
    for i, field in enumerate(t)]
  File "/Users/xiaoli/IdeaProjects/sparkDelivery/python/lib/pyspark.zip/pyspark/serializers.py", line 287, in create_array
    return pa.Array.from_pandas(s, mask=mask, type=t)
  File "pyarrow/array.pxi", line 335, in pyarrow.lib.Array.from_pandas (/Users/travis/build/BryanCutler/arrow-dist/arrow/python/build/temp.macosx-10.6-intel-2.7/lib.cxx:30884)
    return array(obj, mask=mask, type=type, memory_pool=memory_pool,
  File "pyarrow/array.pxi", line 170, in pyarrow.lib.array (/Users/travis/build/BryanCutler/arrow-dist/arrow/python/build/temp.macosx-10.6-intel-2.7/lib.cxx:29224)
    return _ndarray_to_array(values, mask, type, from_pandas, pool)
  File "pyarrow/array.pxi", line 70, in pyarrow.lib._ndarray_to_array (/Users/travis/build/BryanCutler/arrow-dist/arrow/python/build/temp.macosx-10.6-intel-2.7/lib.cxx:28465)
    check_status(NdarrayToArrow(pool, values, mask,
  File "pyarrow/error.pxi", line 85, in pyarrow.lib.check_status (/Users/travis/build/BryanCutler/arrow-dist/arrow/python/build/temp.macosx-10.6-intel-2.7/lib.cxx:8570)
    raise ArrowNotImplementedError(message)
ArrowNotImplementedError: No cast implemented from string to int32

	at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:453)
	at org.apache.spark.sql.execution.python.ArrowPythonRunner$$anon$2.read(ArrowPythonRunner.scala:172)
	at org.apache.spark.sql.execution.python.ArrowPythonRunner$$anon$2.read(ArrowPythonRunner.scala:122)
	at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
	at org.apache.spark.sql.execution.python.ArrowEvalPythonExec$$anon$2.<init>(ArrowEvalPythonExec.scala:102)
	at org.apache.spark.sql.execution.python.ArrowEvalPythonExec.evaluate(ArrowEvalPythonExec.scala:100)
	at org.apache.spark.sql.execution.python.EvalPythonExec.$anonfun$doExecute$2(EvalPythonExec.scala:126)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2(RDD.scala:817)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitions$2$adapted(RDD.scala:817)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:327)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:291)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:327)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:291)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:327)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:291)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:327)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:291)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:121)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:428)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1341)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:431)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
19/03/21 00:04:08 WARN TaskSetManager: Lost task 0.0 in stage 8.0 (TID 8, localhost, executor driver): org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/Users/xiaoli/IdeaProjects/sparkDelivery/python/lib/pyspark.zip/pyspark/worker.py", line 433, in main
    process()
  File "/Users/xiaoli/IdeaProjects/sparkDelivery/python/lib/pyspark.zip/pyspark/worker.py", line 428, in process
    serializer.dump_stream(func(split_index, iterator), outfile)
  File "/Users/xiaoli/IdeaProjects/sparkDelivery/python/lib/pyspark.zip/pyspark/serializers.py", line 360, in dump_stream
    self._assign_cols_by_name)
  File "/Users/xiaoli/IdeaProjects/sparkDelivery/python/lib/pyspark.zip/pyspark/serializers.py", line 316, in _create_batch
    for i, field in enumerate(t)]
  File "/Users/xiaoli/IdeaProjects/sparkDelivery/python/lib/pyspark.zip/pyspark/serializers.py", line 287, in create_array
    return pa.Array.from_pandas(s, mask=mask, type=t)
  File "pyarrow/array.pxi", line 335, in pyarrow.lib.Array.from_pandas (/Users/travis/build/BryanCutler/arrow-dist/arrow/python/build/temp.macosx-10.6-intel-2.7/lib.cxx:30884)
    return array(obj, mask=mask, type=type, memory_pool=memory_pool,
  File "pyarrow/array.pxi", line 170, in pyarrow.lib.array (/Users/travis/build/BryanCutler/arrow-dist/arrow/python/build/temp.macosx-10.6-intel-2.7/lib.cxx:29224)
    return _ndarray_to_array(values, mask, type, from_pandas, pool)
  File "pyarrow/array.pxi", line 70, in pyarrow.lib._ndarray_to_array (/Users/travis/build/BryanCutler/arrow-dist/arrow/python/build/temp.macosx-10.6-intel-2.7/lib.cxx:28465)
    check_status(NdarrayToArrow(pool, values, mask,
  File "pyarrow/error.pxi", line 85, in pyarrow.lib.check_status (/Users/travis/build/BryanCutler/arrow-dist/arrow/python/build/temp.macosx-10.6-intel-2.7/lib.cxx:8570)
    raise ArrowNotImplementedError(message)
```

Reply:
cc @BryanCutler @HyukjinKwon @ueshin

Reply:
Also, is it possible the mismatched types could generate wrong results?

Reply:
Yes, something has to be done. I was at the very least tried to document the casting 
combinations.

**Pandas UDF matrix:**

https://github.com/apache/spark/blob/b67d36957287c1fbefa1996e6a4a009a75c4c3f8/python/pyspark/sql/functions.py#L3098-L3131

The problem is, this matrix is different from regular PySpark UDF, and also our `TypeCoercions`:

**Regular PySpark UDF matrix:**

https://github.com/apache/spark/blob/b67d36957287c1fbefa1996e6a4a009a75c4c3f8/python/pyspark/sql/functions.py#L2830-L2860

I lost the last discussion about whether we should allow such type coercions or not. I roughly recall: If we allow, I think it will need a huge bunch of codes to maintain again (Arrow Type <> Pandas type <> Python type <> SparkSQL type), but if we disallow, it will break many existing apps.

Reply:
And .. yes I think we should also throw better exceptions at the very least.

Reply:
Sometimes type casting is necessary, for example if your integer column has null values then Pandas will automatically upcast to floating point to represent the nulls as NaNs. If that column is returned then it doesn't make sense to keep it as floating point because Spark can handle the null values, so using integer return type will cause type casting, but won't cause any problems.